pkg/dockerregistry/server/localblobstore.go

```diff
@@ -0,0 +1,139 @@
+package server
+
+import (
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+
+	"k8s.io/kubernetes/pkg/api/errors"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type localBlobStore struct {
+	distribution.BlobStore
+
+	repo *repository
+}
+
+var _ distribution.BlobStore = &localBlobStore{}
+
+func (bs *localBlobStore) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
+	desc, err := bs.BlobStore.Stat(ctx, dgst)
+	switch {
+	case err == distribution.ErrBlobUnknown:
+		// continue on to the code below and look up the blob in a local store
+	case err != nil:
+		context.GetLogger(bs.repo.ctx).Errorf("Failed to find blob %s: %#v", dgst.String(), err)
+		fallthrough
+	default:
+		return desc, err
+	}
+
+	is, err := bs.repo.getImageStream()
+	if err != nil {
+		if errors.IsNotFound(err) || errors.IsForbidden(err) {
+			return distribution.Descriptor{}, distribution.ErrBlobUnknown
+		}
+		context.GetLogger(bs.repo.ctx).Errorf("Error retrieving image stream for blob: %s", err)
+		return distribution.Descriptor{}, err
+	}
+
+	cached := bs.repo.cachedLayers.RepositoriesForDigest(dgst)
+
+	// look at the first level of tagged repositories first
+	search := bs.identifyCandidateRepositories(is, true)
```

Comment:
We could construct a new `Repository` object for the desired local repository we want to stat. However, we'd have to once more go through all the layers `server.repository -> server.pullthroughBlobStore -> server.localBlobStore -> ...` which we'd like to eliminate. Maybe we could set something like `"openshift.middleware.locallookup=true"` to *turn them off*.